task_2/join.go

```diff
@@ -0,0 +1,124 @@
+package task_2
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+)
+
+func Start() {
+
+	var (
+		numOfChannels = rand.Intn(12) + 1
+		channels      []chan interface{}
+	)
+
+	fmt.Println(fmt.Sprintf("creating %d channels..", numOfChannels))
+	for i := 0; i < numOfChannels; i++ {
+		channels = append(channels, make(chan interface{}))
+	}
+
+	ctx, _ := context.WithTimeout(context.Background(), 35*time.Second)
+
+	joinedChan := join(ctx, channels...)
+	go writeToChannels(channels)
+	for {
+		select {
+		case <-time.After(10 * time.Second):
+			fmt.Println("reached time limit waiting for joined output, exiting.")
+			return
+		case joinedOutput, ok := <-joinedChan:
+			if !ok {
+				return
+			}
+			fmt.Println("[RECEIVED JOINED OUTPUT]: ", joinedOutput)
+		}
+	}
+
+}
+
+func writeToChannels(ch []chan interface{}) {
+	for i := 1; i < 10; i++ {
+		fmt.Println(fmt.Sprintf("-- CYCLE %d -- sending messages to channels..", i))
+		for j, c := range ch {
+			c <- fmt.Sprintf("<cycle %d hi_%d>", i, j)
+		}
+		time.Sleep(1 * time.Second)
+	}
+}
+
+func join(ctx context.Context, channels ...chan interface{}) chan interface{} {
+	joinedChan := make(chan interface{})
+	beginNewCycleChan := make(chan interface{}, 6)
+	output := make(chan interface{})
+	var wg sync.WaitGroup
+	wg.Add(len(channels))
+	messages := make([]interface{}, 0, len(channels))
+
+	for i, c := range channels {
+		fmt.Println("dispatching goroutine to receive from channel..")
+		go func(ctx context.Context, input <-chan interface{}, beginNewCycleChan chan interface{}, output chan interface{}, index int) {
+			defer wg.Done()
+			var sentMessageThisCycle bool
+			for {
+
+				if !sentMessageThisCycle {
```

Comment:
You can use a trick provided in the book "Concurrency in Go" and adopt it here.
```go
go func(in1, in2 <-chan int) {
	for {
		log.Println("Start new iteration")
		var t1 = in1
		var t2 = in2
		for {
			select {
			case v1 := <-t1:
				log.Println("got from first input", v1)
				t1 = nil
			case v2 := <-t2:
				log.Println("got from second input", v2)
				t2 = nil
			}
			if t1 == nil && t2 == nil {
				break
			}
		}
		log.Println("End iteration")
	}
}(s1, s2)
```
To simplify, we can use a fixed number of joined channels and later evolve to the dynamical number of input channels.

Reply:
For future improvements, please take a look and find more details about `reflect.Select` like [here](https://gist.github.com/17twenty/ada7dd263e9a957126571fa56c8d280f)